developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,255 @@
+
+VirtIO is a platform-agnostic virtualisation standard, managed by OASIS
+https://www.oasis-open.org/committees/tc_home.php?wg_abbrev=virtio . It defines
+a design and a protocol for implementing device IO in VMs.
+
+When Linux is running in a VM, multiple drivers are already available, using
+virtio: networking, storage, etc. Those drivers use the virtio and virtqueue
+APIs to communicate with the host or, potentially, with other guests.
+
+The host part depends on the used virtualisation solution. Some of them
+implement the VirtIO API themselves, then using their method of choice to talk
+to the whatever host they are running on. A typical Linux virtualisation
+implementation uses QEMU + KVM to run guest instances. In this case there are
+two ways to implement a VirtIO host: in the user space and in the kernel. If
+implemented in the user space, it is managed by QEMU. In the kernel space VirtIO
+host drivers use the Linux vhost API.
+
+It is also worth to note, that VirtIO guest drivers reside in respective driver
+subdirectories in the Linux kernel, e.g. drivers/net, drivers/block etc. On the
+contrary, all vhost drivers are located under drivers/vhost. That directory also
+contains vhost.c and vhost.h which are used by all vhost drivers.
+
+The SOF virtualisation solution also implements a vhost driver.
+
+2. SOF VirtIO design
+********************
+
+The principal point of the SOF VirtIO design is using the IPC API, that is also
+used by the native SOF driver to communicate with the SOF code, running on the
+audio DSP. We use the same IPC structures and we add 3 more VirtIO-specific IPC
+messages, which are never sent to the DSP, so the firmware doesn't need to
+support them. For the most part IPCs from guests are forwarded one-to-one by the
+host driver to the DSP.
+
+2.1. Communication channels
+---------------------------
+
+The goal of the SOF virtualisation is to implement low latency audio
+virtualisation, which means avoiding copying of audio data and stream position
+updates. In this case only one VirtQueue would be required for IPC message
+exchange. However, during the first stage we also transfer data and stream
+position updates via VirtQueues. That means, that currently the SOF uses 3
+VirtQueues: for IPC, audio data, and stream position updates.
+
+TODO: even though currently both playback and capture are supported, they aren't
+supported simultaneously. This should be fixed in future versions.
+
+2.2. Topology
+-------------
+
+When running natively, the SOF uses the ALSA topology API to configure audio
+pipelines on the DSP. The SOF driver reads a platform-specific topology file
+from the filesystem, parses it and sends all the components from it to the DSP
+firmware.
+
+When configured to support virtualised audio, the host topology additionally
+provides connection points for guest audio pipelines. Guest playback pipelines
+connect to mixer and capture pipelines to demultiplexer components. The ALSA /
```

Comment:
not necessarily. You may want to dedicate a DAI to a guest. or run completely different topologies. It's not correct to assume the sound from a guest would always be mixed.

Reply:
Would this be ok if I keep references to mixer / demux but explain, that this isn't a requirement, but just the currently used configuration? At least I don't think I have any explicit checks or dependencies for that in the code. I only care about linking component IDs

Reply:
it's really not great to explain as follows:

guest can connect to foobar but foobar is not a requirement and other connections are possible

You may want to explicitly state what the guest can connect to, so that there's no ambiguity for the reader/reviewer.

Reply:
@plbossart but it isn't just one of many possibilities, I think there are 2 options: BE sharing with the host or other guests and dedicated BEs. When sharing you use mixer / demux, when dedicated you use more or less anything else?

Reply:
yes, this description is accurate and should be added

Reply:
@plbossart it's already there in the latest version